### ISSUE TITLE
openssl x509 -req ignores -CAserial/-CAcreateserial options

### ISSUE THREAD

**Issue by arvidjaar** [openssl/openssl#23170]
The following command
```
openssl x509 -req -in example.org.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out example.org.crt
```
is supposed to create serial file `ca.srl`. It does not do it. It was apparently broken in the commit 05458fdb73dcca30edace5ad727a15d6d919e215 which removed `sno = NULL` assignment in `apps/x509.c:x509_main()`:
```
-            if (!X509_set_serialNumber(x, sno))
-                goto end;
-            ASN1_INTEGER_free(sno);
-            sno = NULL;
```
Later if `-CA` option is present, `x509_main` would load serial file if `sno == NULL`, but `sno` is now not NULL at this point.
```
    if (CAfile != NULL) {
        issuer_cert = xca;
        if (sno == NULL)
            sno = x509_load_serial(CAfile, CAserial, CA_createserial);
```
so `x509_load_serial` is never called.

**Comment by t8m** [comment]
You now need to specify both `-CAserial` and `-CAcreateserial` options to have the ca.srl created.

Citing the manual page:
```
If the B<-CA> option is specified and both the <-CAserial> and <-CAcreateserial>
options are not given and the default serial number file does not exist,
a random number is generated; this is the recommended practice.
```

**Comment by arvidjaar** [comment]
> You now need to specify both `-CAserial` and `-CAcreateserial` options to have the ca.srl created.

```
bor@bor-Latitude-E5450:~/tmp/CA$ openssl x509 -req -in dovecot.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out dovecot.crt  -CAserial xxx.srl
Certificate request self-signature ok
subject=C = ES, ST = Murcia, L = Cartagena, O = Valinor, OU = IMAP server, CN = telcontar.valinor, emailAddress = postmaster@telcontar.valinor
bor@bor-Latitude-E5450:~/tmp/CA$ ll
итого 32
drwxrwxr-x  2 bor bor  4096 янв  2 14:14 ./
drwxrwxr-x 19 bor bor 12288 янв  1 19:02 ../
-rw-rw-r--  1 bor bor  1245 дек 30 08:59 ca.crt
-rw-------  1 bor bor  1704 дек 30 08:59 ca.key
-rw-rw-r--  1 bor bor  1074 янв  2 14:15 dovecot.crt
-rw-rw-r--  1 bor bor   785 янв  2 14:14 dovecot.csr
bor@bor-Latitude-E5450:~/tmp/CA$ openssl version
OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
bor@bor-Latitude-E5450:~/tmp/CA$ 
```

**Comment by t8m** [comment]
Ah that bug was fixed after the 3.0.2 release.

**Comment by arvidjaar** [comment]
You are right, it is fixed in 3.1.4 and master. Sorry for the noise. Commit message is misleading, `-CAcreateserial` does not need `-CAserial` option.

```
bor@tw:~/ssl> ll ca.*
-rw-r--r-- 1 bor bor 1245 Jan  2 16:09 ca.crt
-rw------- 1 bor bor 1704 Jan  2 16:09 ca.key
bor@tw:~/ssl> openssl x509 -req -in example.org.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out example.org.crt
Certificate request self-signature ok
subject=C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
bor@tw:~/ssl> ll ca.*
-rw-r--r-- 1 bor bor 1245 Jan  2 16:09 ca.crt
-rw------- 1 bor bor 1704 Jan  2 16:09 ca.key
-rw-r--r-- 1 bor bor   41 Jan  2 16:20 ca.srl
bor@tw:~/ssl> openssl version
OpenSSL 3.1.4 24 Oct 2023 (Library: OpenSSL 3.1.4 24 Oct 2023)
bor@tw:~/ssl> 
```
